mod/beacon/blockchain/service.go

```diff
@@ -164,134 +166,154 @@ func (s *Service[
-			if msg.Type() == events.GenesisDataProcessRequest {
-				s.handleProcessGenesisDataRequest(msg)
-			}
+		case event := <-s.subGenDataReceived:
+			s.handleGenDataReceived(event)
+		case event := <-s.subBlockReceived:
+			s.handleBeaconBlockReceived(event)
+		case event := <-s.subFinalBlkReceived:
+			s.handleFinalBeaconBlockReceived(event)
 		}
 	}
 }
 
+/* -------------------------------------------------------------------------- */
+/*                              Message Handlers                              */
+/* -------------------------------------------------------------------------- */
+
 func (s *Service[
 	_, _, _, _, _, _, _, _, GenesisT, _,
-]) handleProcessGenesisDataRequest(msg *asynctypes.Event[GenesisT]) {
+]) handleGenDataReceived(msg events.Event[GenesisT]) {
+	var (
+		valUpdates transition.ValidatorUpdates
+		genesisErr error
+	)
 	if msg.Error() != nil {
 		s.logger.Error("Error processing genesis data", "error", msg.Error())
-		return
 	}
 
 	// Process the genesis data.
-	valUpdates, err := s.ProcessGenesisData(msg.Context(), msg.Data())
-	if err != nil {
-		s.logger.Error("Failed to process genesis data", "error", err)
+	valUpdates, genesisErr = s.ProcessGenesisData(msg.Context(), msg.Data())
+	if genesisErr != nil {
+		s.logger.Error("Failed to process genesis data", "error", genesisErr)
 	}
 
-	// Publish the validator set updated event.
-	if err = s.validatorUpdateBroker.Publish(
-		msg.Context(),
-		asynctypes.NewEvent(
+	// Emit the event containing the validator updates.
+	if err := s.dispatcher.Publish(
+		events.New(
 			msg.Context(),
-			events.ValidatorSetUpdated,
+			events.GenesisDataProcessed,
 			valUpdates,
-			err,
+			genesisErr,
 		),
 	); err != nil {
 		s.logger.Error(
-			"Failed to publish validator set updated event",
-			"error",
-			err,
+			"Failed to emit event in process genesis data",
+			"error", err,
 		)
+		panic(err)
 	}
 }
 
 func (s *Service[
 	_, BeaconBlockT, _, _, _, _, _, _, _, _,
 ]) handleBeaconBlockReceived(
-	msg *asynctypes.Event[BeaconBlockT],
+	msg events.Event[BeaconBlockT],
 ) {
 	// If the block is nil, exit early.
 	if msg.Error() != nil {
 		s.logger.Error("Error processing beacon block", "error", msg.Error())
 		return
 	}
 
-	// Publish the verified block event.
-	if err := s.blkBroker.Publish(
-		msg.Context(),
-		asynctypes.NewEvent(
+	// emit a BeaconBlockVerified event with the error result from \
+	// VerifyIncomingBlock
+	if err := s.dispatcher.Publish(
+		events.New(
 			msg.Context(),
 			events.BeaconBlockVerified,
 			msg.Data(),
 			s.VerifyIncomingBlock(msg.Context(), msg.Data()),
 		),
 	); err != nil {
-		s.logger.Error("Failed to publish verified block", "error", err)
+		s.logger.Error(
+			"Failed to emit event in verify beacon block",
+			"error", err,
+		)
 	}
 }
 
 func (s *Service[
 	_, BeaconBlockT, _, _, _, _, _, _, _, _,
-]) handleBeaconBlockFinalization(
-	msg *asynctypes.Event[BeaconBlockT],
+]) handleFinalBeaconBlockReceived(
```

Comment:
I would revert this, since the fact that we are "receiving" the block again from CometBFT is a CometBFT specific thing, in a lot of other consensus engines, we might just load a stored beacon block, that was stored during verifiy.